<a href="https://colab.research.google.com/github/alehunter/CAP4770/blob/main/labs/Alejandro_Hunter_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Pandas

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/pandas-logo-2.png)

<font size='0.85em'>(logo by [Marc Garcia](https://github.com/pandas-dev/pandas/blob/master/web/pandas/static/img/pandas.svg))</font> 


> A suggestion: This worksheet is organized in a way that, hopefully, best presents the topics and not in a way that is the most useful for reference. As you progress through this worksheet, your future you will find it useful if you create your own notes so that you can quickly find the relevant information. These notes might take the form of a separate Google Colab notebook or handwritten notes. In addition, just the activity of taking notes will help you retain the information.

The Pandas Library is built on top of Numpy and is designed to make working with data fast and easy. Like Numpy, the library includes data structures and functions to manipulate that data.

As we learned in the Numpy Notebook, we need to load in the library before we can use it.

In [153]:
from pandas import Series, DataFrame
import numpy as np
import pandas as pd

Let's dissect the code above.

The `Series` and `DataFrame` datatypes are commonly used so we import them directly with

```
from pandas import Series, DataFrame
```

For all other datatypes and functions in the library, the `pd` prefix is commonly used so we import that with

```
import pandas as pd
```

## Series
A series is a 1d array-like object

Let's consider the heights (in cm) of the members of the Japan's Women's Basketball Team at the 2020 Olympics (they won the silver medal).      

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/basketball2020.png)
We can create a series in a number of ways.

For example, we can create it directly from a Python list

In [154]:
japan = [183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182] # a Python list
athletesHeight = Series(japan) # converted to a Pandas Series

We could also have done

In [155]:
athletesHeight = Series([183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182])

In either case we can see the value of the Series `athletesHeight`

In [156]:
 athletesHeight

0     183
1     185
2     167
3     162
4     165
5     174
6     173
7     181
8     167
9     183
10    185
11    182
dtype: int64

You are probably familiar with arrays in programming languages and this works in a similar way. The left number is the index and the right the value. And we can find the value at a particular index by the usual:

In [157]:
athletesHeight[3]

162

#### specifying indices
Instead of the index 0, 1, 2, 3 ... you can specify your own index values. For example, we can label them 'Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi' ... etc. 

In [158]:
athletes2 = Series(japan, index = ['Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi', 
                                   'Rui Machida', 'Nako Motohashi', 'Nanaka Todo', 
                                   'Saki Hayashi','Evelyn Mawuli', 'Saori Miyazaki', 'Yuki Miyazawa', 'Himawari Akaho', 
                                   'Monica Okoye'])
athletes2

Moeko Nagaoka     183
Maki Takada       185
Naho Miyoshi      167
Rui Machida       162
Nako Motohashi    165
Nanaka Todo       174
Saki Hayashi      173
Evelyn Mawuli     181
Saori Miyazaki    167
Yuki Miyazawa     183
Himawari Akaho    185
Monica Okoye      182
dtype: int64

The names we see are not another column of the data. We can see the shape of athletes2 by:

In [159]:
athletes2.shape

(12,)

This shows that athletes2 is a one dimensional matrix and that dimension has a length of 12. So the names we see are not values in a column but rather the indices.


Let's use the index to get the Height of Himawari Akaho:

In [160]:
athletes2['Himawari Akaho']

185


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)
## DataFrame
DataFrames are **the most important data structure of Pandas** and are simply
a table or spreadsheet like structure.  A DataFrame represents a table like:
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

### Creating a DataFrame by Hand

A common way to create a DataFrame is to use a Python dictionary as follows:

In [161]:
 cars = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [256, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, 113, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}


and now we can create a DataFrame from the `cars` Python dictionary:

In [162]:
df = DataFrame(cars)
df


Make Drive  Range  Battery_kwH  PeakPower_kW  MPGe  0-60
0         Ioniq 5   AWD    256        77.40           239    98  5.00
1   Tesla Model 3   AWD    315        80.00           298   113  3.10
2      Rivian R1T   AWD    316       135.00           562    70  3.00
3         VW ID.4   RWD    260        82.00           260   107  7.70
4  F150 Lightning   AWD    230       110.00           318    68  4.00

Prior to my life with Pandas, I would represent a table like the above one as:

In [163]:
prePandas = [{'Make': 'Ioniq 5', 'Drive': 'AWD', 'Range': 256, 'Battery_kwH': 77.4, 'PeakPower_kW': 98, 'MPGe': 98, '0-60': 5.0},
             {'Make': 'Tesla Model 3', 'Drive': 'AWD', 'Range': 315, 'Battery_kwH': 80.0, 'PeakPower_kW': 113,'MPGe': 113, '0-60': 3.1},
             {'Make': 'Rivian R1T', 'Drive': 'AWD', 'Range': 316, 'Battery_kwH': 135.0, 'PeakPower_kW': 562,'MPGe': 70, '0-60': 3.0},
             {'Make': 'VW ID.4', 'Drive': 'RWD', 'Range': 260, 'Battery_kwH': 82.0, 'PeakPower_kW': 260,'MPGe': 107, '0-60': 7.7},
             {'Make': 'F150 Lightning', 'Drive': 'AWD', 'Range': 230, 'Battery_kwH': 110.0, 'PeakPower_kW': 318,'MPGe': 68, '0-60': 4.0}]


prePandas[0]['Make']

'Ioniq 5'

In the prePandas scheme the data is organized first by rows. That seemed logical to me since each row represents an object and is how we organize data in an SQL database. In the Pandas representation the data is organized by columns. 



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Reading data from a csv file:

A CSV file (comma separated values) is a common data format for data science files. As| the name suggests, the data values in a row are separated by commas. For example, the data file to represent the table:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0


would be

```
Make,Drive,Range,Battery_kwH,PeakPower_kW,MPGe,0-60
Ioniq 5,AWD,256,77.4,239,98,5.0
Tesla Model 3,AWD,315,80,298,113,3.1
Rivian R1T,AWD,316,135,562,70,3.0 
VW ID.4,RWD,260,82,260,107,7.7
F150 Lightning,AWD,230,110,318,68,4.0

```


As the name suggests, we use the `pd.read_csv` function to read a csv file.  `pd.read_csv` can read a csv file from either your local machine or the web. Let's start with the web.

### Reading a CSV file from the web.
To read a file from the web, we simply provide a URL:


In [164]:
evs = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv')
evs

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor      4.60           233   
1    Volkswagen                        ID.3 Pure     10.00           160   
2      Polestar                                2      4.70           210   
3           BMW                              iX3      6.80           180   
4         Honda                                e      9.50           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance      5.10           200   
102       Byton                M-Byte 95 kWh 2WD      7.50           190   
103      Rivian                              R1T      3.00           185   
104         GMC                        Hummer EV      3.00           171   
105        Ford                   F150 Lightning      3.00           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Sometimes the csv file has a header row as was the case in the example above. That file starts with the line
```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
```
as we can see by using `curl`



In [165]:
!curl -s https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv |  head

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
Tesla,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
BMW,iX3,6.8,180,360,206,560,Yes,RWD,Type 2 CCS,SUV,D,5,68040
Honda,e,9.5,145,170,168,190,Yes,RWD,Type 2 CCS,Hatchback,B,4,32997
Lucid,Air,2.8,250,610,180,620,Yes,AWD,Type 2 CCS,Sedan,F,5,105000
Volkswagen,e-Golf,9.6,150,190,168,220,Yes,FWD,Type 2 CCS,Hatchback,C,5,31900
Peugeot,e-208,8.1,150,275,164,420,Yes,FWD,Type 2 CCS,Hatchback,B,5,29682
Tesla,Model 3 Standard Range Plus,5.6,225,310,153,650,Yes,RWD,Type 2 CCS,Sedan,D,5,46380


**An aside**

We can preface any Unix command with a bang (!) to have it execute in our Notebook.  This is amazingly handy:



In [166]:
!ls

data  sample_data


In [167]:
!pwd

/content


### Files with No Header Row
Sometimes the csv file does not have a header row. So for example, data might start on the very first line of the file

```
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen ,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar ,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
```

In that case you specify the names of the columns using the `names` parameter:

In [168]:
columnNames = ['Brand','Model','AccelSec','TopSpeed_KmH','Range_Km','Efficiency_WhKm',
               'FastCharge_KmH','RapidCharge','PowerTrain','PlugType','BodyStyle',
               'Segment','Seats','PriceEuro']

evs2 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean_NoHeader.csv', names=columnNames)
evs2

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla     Model 3 Long Range Dual Motor      4.60           233   
1    Volkswagen                         ID.3 Pure     10.00           160   
2      Polestar                                 2      4.70           210   
3           BMW                              iX3       6.80           180   
4         Honda                                e       9.50           145   
..           ...                              ...       ...           ...   
98       Nissan                       Ariya 63kWh      7.50           160   
99         Audi     e-tron S Sportback 55 quattro      4.50           210   
100      Nissan               Ariya e-4ORCE 63kWh      5.90           200   
101      Nissan   Ariya e-4ORCE 87kWh Performance      5.10           200   
102       Byton                 M-Byte 95 kWh 2WD      7.50           190   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
98        330              191            440         Yes        FWD   
99        335              258            540         Yes        AWD   
100       325              194            440         Yes        AWD   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5      55480  
1    Type 2 CCS  Hatchback       C      5      30000  
2    Type 2 CCS   Liftback       D      5      56440  
3    Type 2 CCS        SUV       D      5      68040  
4    Type 2 CCS  Hatchback       B      4      32997  
..          ...        ...     ...    ...        ...  
98   Type 2 CCS  Hatchback       C      5      45000  
99   Type 2 CCS        SUV       E      5      96050  
100  Type 2 CCS  Hatchback       C      5      50000  
101  Type 2 CCS  Hatchback       C      5      65000  
102  Type 2 CCS        SUV       E      5      62000  

[103 rows x 14 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

### Reading a CSV file the local machine
First, let's get that file onto our local machine:

In [169]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv > evs.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  8268  100  8268    0     0   171k      0 --:--:-- --:--:-- --:--:--  175k


Hmm. That didn't work. Can you fix the error and rerun that cell?

---
Now we can specify the local file using `pd=read_csv`

In [170]:
ev3 = pd.read_csv('evs.csv')
ev3

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor      4.60           233   
1    Volkswagen                        ID.3 Pure     10.00           160   
2      Polestar                                2      4.70           210   
3           BMW                              iX3      6.80           180   
4         Honda                                e      9.50           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance      5.10           200   
102       Byton                M-Byte 95 kWh 2WD      7.50           190   
103      Rivian                              R1T      3.00           185   
104         GMC                        Hummer EV      3.00           171   
105        Ford                   F150 Lightning      3.00           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Suppose we want that file in a data directory. Let's go ahead and create the directory and move the file there.

In [171]:
!mkdir data
!mv evs.csv data

mkdir: cannot create directory ‘data’: File exists


Now when we load the file we need to give more of a path:

In [172]:
evs4 = pd.read_csv('data/evs.csv')
evs4

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor      4.60           233   
1    Volkswagen                        ID.3 Pure     10.00           160   
2      Polestar                                2      4.70           210   
3           BMW                              iX3      6.80           180   
4         Honda                                e      9.50           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance      5.10           200   
102       Byton                M-Byte 95 kWh 2WD      7.50           190   
103      Rivian                              R1T      3.00           185   
104         GMC                        Hummer EV      3.00           171   
105        Ford                   F150 Lightning      3.00           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Missing Data


In real machine learning tasks, we often encounter missing values.

### Missing Data in Files
For example, suppose we didn't know The Polestar 2's acceleration, the BMW iX3's top speed and the Honda's model name. In that case our CSV file would start


```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,,210,400
BMW ,iX3 ,6.8,,360
Honda ,,9.5,145,
```

with the double comma on the Polestar, BMW and Honda lines representing the missing data. When we read that file.



In [173]:
evs5 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small.csv')
evs5

Brand                          Model  AccelSec  TopSpeed_KmH  \
0        Tesla   Model 3 Long Range Dual Motor      4.60        233.00   
1   Volkswagen                       ID.3 Pure     10.00        160.00   
2     Polestar                               2       NaN        210.00   
3          BMW                            iX3       6.80           NaN   
4        Honda                             NaN      9.50        145.00   
5        Lucid                            Air       2.80        250.00   
6   Volkswagen                         e-Golf       9.60        150.00   
7      Peugeot                          e-208       8.10        150.00   
8        Tesla     Model 3 Standard Range Plus      5.60        225.00   
9         Audi                      Q4 e-tron       6.30        180.00   
10    Mercedes                  EQC 400 4MATIC      5.10        180.00   
11      Nissan                           Leaf       7.90        144.00   
12     Hyundai            Kona Electric 64 kWh      7.90        167.00   
13         BMW                             i4       4.00        200.00   
14     Hyundai                  IONIQ Electric      9.70        165.00   
15  Volkswagen                      ID.3 Pro S      7.90        160.00   
16     Porsche                  Taycan Turbo S      2.80        260.00   
17  Volkswagen                          e-Up!      11.90        130.00   
18          MG                           ZS EV      8.20        140.00   
19        Mini                      Cooper SE       7.30        150.00   

    Range_Km  
0     450.00  
1     270.00  
2     400.00  
3     360.00  
4        NaN  
5     610.00  
6     190.00  
7     275.00  
8     310.00  
9     400.00  
10    370.00  
11    220.00  
12    400.00  
13    450.00  
14    250.00  
15    440.00  
16    375.00  
17    195.00  
18    220.00  
19    185.00

We see that missing values are now the floating point values NaN meaning Not a Number. This NaN value is used even in columns that do not contain floating point values. For example, in row 4 above, Honda has NaN in the model name column. Wes McKinney, the developer of Pandas calls NaN a *sentinel* value that is easily detected and indicates a missing value.

### Special Characters Representing NaN in the Data File

Sometimes, special characters are used in a data file to represent missing values. For example, sometimes a dash is used ...

Brand|Model|AccelSec|TopSpeed_KmH|Range_Km
:--- | :--- | ---: | ---: | ---: |
Tesla |Model 3 Long Range Dual Motor|4.6|233|450
Volkswagen |ID.3 Pure|10.0|160|270
Polestar |2|-|210|400
BMW |iX3 |6.8|-|360
Honda |-|9.5|145|

The associated data file would look like

```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,-,210,400
BMW ,iX3 ,6.8,-,360
Honda ,-,9.5,145,
```

If we do not convert these to `NaN`, these dashes will create havoc with future calculations:

``` 
TypeError: can't multiply sequence by non-int of type 'float'
```

When we read in the csv file we need to convert the dashes to `NaN` by using the `na_values` parameter in `read_csv`. As you can see in the following example, our data file has the dashes 

In [174]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,-,210,400
BMW ,iX3 ,6.8,-,360
Honda ,-,9.5,145,
Lucid ,Air ,2.8,250,610
Volkswagen ,e-Golf ,9.6,150,190
Peugeot ,e-208 ,8.1,150,275
Tesla ,Model 3 Standard Range Plus,5.6,225,310
Audi ,Q4 e-tron ,6.3,180,400
Mercedes ,EQC 400 4MATIC,5.1,180,370
Nissan ,Leaf ,7.9,144,220
Hyundai ,Kona Electric 64 kWh,7.9,167,400
BMW ,i4 ,4.0,200,450
Hyundai ,IONIQ Electric,9.7,165,250
Volkswagen ,ID.3 Pro S,7.9,160,440
Porsche ,Taycan Turbo S,2.8,260,375
Volkswagen ,e-Up! ,11.9,130,195
MG ,ZS EV,8.2,140,220
Mini ,Cooper SE ,7.3,150,185


and we can convert those to `NaN` when we read the file

In [175]:
evdash = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv',
                     na_values='-')
evdash[:6]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla   Model 3 Long Range Dual Motor      4.60        233.00   
1  Volkswagen                       ID.3 Pure     10.00        160.00   
2    Polestar                               2       NaN        210.00   
3         BMW                            iX3       6.80           NaN   
4       Honda                             NaN      9.50        145.00   
5       Lucid                            Air       2.80        250.00   

   Range_Km  
0    450.00  
1    270.00  
2    400.00  
3    360.00  
4       NaN  
5    610.00

### Specifying Missing Values by hand
Suppose we didn't know the range of the Hyundai Ioniq 5 and the MPGe of the Tesla.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

In that case we can create a dataframe like:

In [176]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [np.nan, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, np.nan, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN        77.40           239  98.00  5.00
1   Tesla Model 3   AWD 315.00        80.00           298    NaN  3.10
2      Rivian R1T   AWD 316.00       135.00           562  70.00  3.00
3         VW ID.4   RWD 260.00        82.00           260 107.00  7.70
4  F150 Lightning   AWD 230.00       110.00           318  68.00  4.00

where `np.nan` is Numpy's NaN.  We can also use Python's `None`:


In [177]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [None, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, None, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN        77.40           239  98.00  5.00
1   Tesla Model 3   AWD 315.00        80.00           298    NaN  3.10
2      Rivian R1T   AWD 316.00       135.00           562  70.00  3.00
3         VW ID.4   RWD 260.00        82.00           260 107.00  7.70
4  F150 Lightning   AWD 230.00       110.00           318  68.00  4.00

In addition to reading CSV files, there are many other ways of reading in data including from SQL databases, mongoDB, and webpages. See the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/) for details.



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Indices
In all the examples above, when we created a DataFrame, an auto-generated, auto-incrementing index was generated. So our initial data might look like:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

But the DataFrame looks like ...

In [178]:
carz
   

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN        77.40           239  98.00  5.00
1   Tesla Model 3   AWD 315.00        80.00           298    NaN  3.10
2      Rivian R1T   AWD 316.00       135.00           562  70.00  3.00
3         VW ID.4   RWD 260.00        82.00           260 107.00  7.70
4  F150 Lightning   AWD 230.00       110.00           318  68.00  4.00

with the 0, 1, 2 ... indices added (the column on the left). This makes it handy when we want to access a particular row.

In [179]:
carz.loc[4]

Make            F150 Lightning
Drive                      AWD
Range                   230.00
Battery_kwH             110.00
PeakPower_kW               318
MPGe                     68.00
0-60                      4.00
Name: 4, dtype: object

As you can see in the above code, we access a particular row (or rows) with the `loc` method.

Sometimes the rows in our data already have a unique identifier. For example in a data file for U.S. states each state might have the unique 2 character state code


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/grayStates.png)

State|Name|Pop2022|Pop2021|Pop2010|TotalArea|LandArea
:--- | :--- | ---: | ---: | ---: | ---: | ---: | 
AK| Alaska| 720763|724357|713910|665384|570640.95
AL| Alabama|4949697|4934193|4785437|52420|50645.33
AR| Arkansas|3042017|3033946|2921964|53179|52035.48
AZ| Arizona|7640796|7520103|6407172|113990|113594.08

In that case we can use the unique identifier, in this case the 2 letter state code, as our index. 

In [180]:
states = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/usStates.csv',
                     index_col='State')
states

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                                   
AK             Alaska    720763    724357    713910     665384 570640.95
AL            Alabama   4949697   4934193   4785437      52420  50645.33
AR           Arkansas   3042017   3033946   2921964      53179  52035.48
AZ            Arizona   7640796   7520103   6407172     113990 113594.08
CA         California  39664128  39613493  37319502     163695 155779.22
CO           Colorado   5961083   5893634   5047349     104094 103641.89
CT        Connecticut   3546588   3552821   3579114       5543   4842.36
DE           Delaware    998619    990334    899593       2489   1948.54
FL            Florida  22177997  21944577  18845537      65758  53624.76
GA            Georgia  10936299  10830007   9711881      59425  57513.49
HI             Hawaii   1401709   1406430   1363963      10932   6422.63
IA               Iowa   3174426   3167974   3050745      56273  55857.13
ID              Idaho   1896652   1860123   1570746      83569  82643.12
IL           Illinois  12518071  12569321  12840503      57914  55518.93
IN            Indiana   6842385   6805663   6490432      36420  35826.11
KS             Kansas   2919179   2917224   2858190      82278  81758.72
KY           Kentucky   4487233   4480713   4348181      40408  39486.34
LA          Louisiana   4616106   4627002   4544532      52378  43293.90
MA      Massachusetts   6922107   6912239   6566307      10554   7800.06
MD           Maryland   6075314   6065436   5788645      12406   9707.24
ME              Maine   1359677   1354522   1327629      35380  30842.92
MI           Michigan   9995212   9992427   9877510      96714  56538.90
MN          Minnesota   5739781   5706398   5310828      86936  79626.74
MO           Missouri   6184843   6169038   5995974      69707  68741.52
MS        Mississippi   2961536   2966407   2970548      48432  43923.27
MT            Montana   1093117   1085004    990697     147040 145545.80
NC     North Carolina  10807491  10701022   9574323      53819  48617.91
ND       North Dakota    774008    770026    674715      70698  69000.80
NE           Nebraska   1960790   1951996   1829542      77348  76824.17
NH      New Hampshire   1378449   1372203   1316762       9349   8952.65
NJ         New Jersey   8870685   8874520   8799446       8723   7354.22
NM         New Mexico   2109093   2105005   2064552     121590 121298.15
NV             Nevada   3238601   3185786   2702405     110572 109781.18
NY           New York  19223191  19299981  19399878      54555  47126.40
OH               Ohio  11727377  11714618  11539336      44826  40860.69
OK           Oklahoma   4007179   3990443   3759944      69899  68594.92
OR             Oregon   4325290   4289439   3837491      98379  95988.01
PA       Pennsylvania  12805190  12804123  12711160      46054  44742.70
RI       Rhode Island   1062583   1061509   1053959       1545   1033.81
SC     South Carolina   5342388   5277830   4635649      32020  30060.70
SD       South Dakota    902542    896581    816166      77116  75811.00
TN          Tennessee   7001803   6944260   6355311      42114  41234.90
TX              Texas  30097526  29730311  25241971     268596 261231.71
UT               Utah   3363182   3310774   2775332      84897  82169.62
VA           Virginia   8638218   8603985   8023699      42775  39490.09
VT            Vermont    622882    623251    625879       9616   9216.66
WA         Washington   7887965   7796941   6742830      71298  66455.52
WI          Wisconsin   5867518   5852490   5690475      65496  54157.80
WV      West Virginia   1755715   1767859   1854239      24230  24038.21
WY            Wyoming    582233    581075    564487      97813  97093.14

We can access a particular row, say the info for New Mexico by 

In [181]:
states.loc['NM']

Name         New Mexico
Pop2022         2109093
Pop2021         2105005
Pop2010         2064552
TotalArea        121590
LandArea      121298.15
Name: NM, dtype: object

You can access a range of rows by the standard Python method:

In [182]:
states[:5]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                               
AK         Alaska    720763    724357    713910     665384 570640.95
AL        Alabama   4949697   4934193   4785437      52420  50645.33
AR       Arkansas   3042017   3033946   2921964      53179  52035.48
AZ        Arizona   7640796   7520103   6407172     113990 113594.08
CA     California  39664128  39613493  37319502     163695 155779.22

or


In [183]:
states[11:16]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IA         Iowa   3174426   3167974   3050745      56273  55857.13
ID        Idaho   1896652   1860123   1570746      83569  82643.12
IL     Illinois  12518071  12569321  12840503      57914  55518.93
IN      Indiana   6842385   6805663   6490432      36420  35826.11
KS       Kansas   2919179   2917224   2858190      82278  81758.72

or using these index values:

In [184]:
states['IA': 'KS']

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IA         Iowa   3174426   3167974   3050745      56273  55857.13
ID        Idaho   1896652   1860123   1570746      83569  82643.12
IL     Illinois  12518071  12569321  12840503      57914  55518.93
IN      Indiana   6842385   6805663   6490432      36420  35826.11
KS       Kansas   2919179   2917224   2858190      82278  81758.72


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Retrieving particular columns
We can get a column of a DataFrame by using the column name:

In [185]:
states[['Pop2022']]

Pop2022
State          
AK       720763
AL      4949697
AR      3042017
AZ      7640796
CA     39664128
CO      5961083
CT      3546588
DE       998619
FL     22177997
GA     10936299
HI      1401709
IA      3174426
ID      1896652
IL     12518071
IN      6842385
KS      2919179
KY      4487233
LA      4616106
MA      6922107
MD      6075314
ME      1359677
MI      9995212
MN      5739781
MO      6184843
MS      2961536
MT      1093117
NC     10807491
ND       774008
NE      1960790
NH      1378449
NJ      8870685
NM      2109093
NV      3238601
NY     19223191
OH     11727377
OK      4007179
OR      4325290
PA     12805190
RI      1062583
SC      5342388
SD       902542
TN      7001803
TX     30097526
UT      3363182
VA      8638218
VT       622882
WA      7887965
WI      5867518
WV      1755715
WY       582233

and we can get multiple columns by passing a list of column names

In [186]:
states[['Pop2022', 'Pop2010']]

Pop2022   Pop2010
State                    
AK       720763    713910
AL      4949697   4785437
AR      3042017   2921964
AZ      7640796   6407172
CA     39664128  37319502
CO      5961083   5047349
CT      3546588   3579114
DE       998619    899593
FL     22177997  18845537
GA     10936299   9711881
HI      1401709   1363963
IA      3174426   3050745
ID      1896652   1570746
IL     12518071  12840503
IN      6842385   6490432
KS      2919179   2858190
KY      4487233   4348181
LA      4616106   4544532
MA      6922107   6566307
MD      6075314   5788645
ME      1359677   1327629
MI      9995212   9877510
MN      5739781   5310828
MO      6184843   5995974
MS      2961536   2970548
MT      1093117    990697
NC     10807491   9574323
ND       774008    674715
NE      1960790   1829542
NH      1378449   1316762
NJ      8870685   8799446
NM      2109093   2064552
NV      3238601   2702405
NY     19223191  19399878
OH     11727377  11539336
OK      4007179   3759944
OR      4325290   3837491
PA     12805190  12711160
RI      1062583   1053959
SC      5342388   4635649
SD       902542    816166
TN      7001803   6355311
TX     30097526  25241971
UT      3363182   2775332
VA      8638218   8023699
VT       622882    625879
WA      7887965   6742830
WI      5867518   5690475
WV      1755715   1854239
WY       582233    564487

### Returning rows with loc

As we have seen we can retrieve rows using the loc function. 

We can also get rows that match a specific criterion. For example, all states whose 2022 population exceeded 10 million. 

In [187]:
populousStates = states.loc[states['Pop2022'] > 10000000] 
populousStates

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                                   
CA         California  39664128  39613493  37319502     163695 155779.22
FL            Florida  22177997  21944577  18845537      65758  53624.76
GA            Georgia  10936299  10830007   9711881      59425  57513.49
IL           Illinois  12518071  12569321  12840503      57914  55518.93
NC     North Carolina  10807491  10701022   9574323      53819  48617.91
NY           New York  19223191  19299981  19399878      54555  47126.40
OH               Ohio  11727377  11714618  11539336      44826  40860.69
PA       Pennsylvania  12805190  12804123  12711160      46054  44742.70
TX              Texas  30097526  29730311  25241971     268596 261231.71

Or let's say we are interested in the states that lost population between 2021 and 2022.

In [188]:
states.loc[states['Pop2022'] < states['Pop2021']]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                                  
AK            Alaska    720763    724357    713910     665384 570640.95
CT       Connecticut   3546588   3552821   3579114       5543   4842.36
HI            Hawaii   1401709   1406430   1363963      10932   6422.63
IL          Illinois  12518071  12569321  12840503      57914  55518.93
LA         Louisiana   4616106   4627002   4544532      52378  43293.90
MS       Mississippi   2961536   2966407   2970548      48432  43923.27
NJ        New Jersey   8870685   8874520   8799446       8723   7354.22
NY          New York  19223191  19299981  19399878      54555  47126.40
VT           Vermont    622882    623251    625879       9616   9216.66
WV     West Virginia   1755715   1767859   1854239      24230  24038.21

States that are over 10 million population and have lost population in 2022:

In [189]:
states.loc[(states['Pop2022'] < states['Pop2021']) & (states['Pop2022'] > 10000000)]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IL     Illinois  12518071  12569321  12840503      57914  55518.93
NY     New York  19223191  19299981  19399878      54555  47126.40


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Creating new columns

#### State Density

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/stateDensity.png)

Let's say we want to add a new column, Density, that gives the number of people per square mile of land area. So the Formula is



$$Density=\frac{Pop2022}{LandArea}$$

We can do that with


In [190]:
states['Density'] = states['Pop2022'] / states['LandArea']
states[:5][['Name', 'Density']]

Name  Density
State                     
AK         Alaska     1.26
AL        Alabama    97.73
AR       Arkansas    58.46
AZ        Arizona    67.26
CA     California   254.62


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Sorting

Density is nice but it would be nice if we ordered the DataFrame by Density...

In [191]:
states.sort_values(by=['Density'])


Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea  \
State                                                                      
AK             Alaska    720763    724357    713910     665384 570640.95   
WY            Wyoming    582233    581075    564487      97813  97093.14   
MT            Montana   1093117   1085004    990697     147040 145545.80   
ND       North Dakota    774008    770026    674715      70698  69000.80   
SD       South Dakota    902542    896581    816166      77116  75811.00   
NM         New Mexico   2109093   2105005   2064552     121590 121298.15   
ID              Idaho   1896652   1860123   1570746      83569  82643.12   
NE           Nebraska   1960790   1951996   1829542      77348  76824.17   
NV             Nevada   3238601   3185786   2702405     110572 109781.18   
KS             Kansas   2919179   2917224   2858190      82278  81758.72   
UT               Utah   3363182   3310774   2775332      84897  82169.62   
ME              Maine   1359677   1354522   1327629      35380  30842.92   
OR             Oregon   4325290   4289439   3837491      98379  95988.01   
IA               Iowa   3174426   3167974   3050745      56273  55857.13   
CO           Colorado   5961083   5893634   5047349     104094 103641.89   
OK           Oklahoma   4007179   3990443   3759944      69899  68594.92   
AR           Arkansas   3042017   3033946   2921964      53179  52035.48   
AZ            Arizona   7640796   7520103   6407172     113990 113594.08   
MS        Mississippi   2961536   2966407   2970548      48432  43923.27   
VT            Vermont    622882    623251    625879       9616   9216.66   
MN          Minnesota   5739781   5706398   5310828      86936  79626.74   
WV      West Virginia   1755715   1767859   1854239      24230  24038.21   
MO           Missouri   6184843   6169038   5995974      69707  68741.52   
AL            Alabama   4949697   4934193   4785437      52420  50645.33   
LA          Louisiana   4616106   4627002   4544532      52378  43293.90   
WI          Wisconsin   5867518   5852490   5690475      65496  54157.80   
KY           Kentucky   4487233   4480713   4348181      40408  39486.34   
TX              Texas  30097526  29730311  25241971     268596 261231.71   
WA         Washington   7887965   7796941   6742830      71298  66455.52   
NH      New Hampshire   1378449   1372203   1316762       9349   8952.65   
TN          Tennessee   7001803   6944260   6355311      42114  41234.90   
MI           Michigan   9995212   9992427   9877510      96714  56538.90   
SC     South Carolina   5342388   5277830   4635649      32020  30060.70   
GA            Georgia  10936299  10830007   9711881      59425  57513.49   
IN            Indiana   6842385   6805663   6490432      36420  35826.11   
HI             Hawaii   1401709   1406430   1363963      10932   6422.63   
VA           Virginia   8638218   8603985   8023699      42775  39490.09   
NC     North Carolina  10807491  10701022   9574323      53819  48617.91   
IL           Illinois  12518071  12569321  12840503      57914  55518.93   
CA         California  39664128  39613493  37319502     163695 155779.22   
PA       Pennsylvania  12805190  12804123  12711160      46054  44742.70   
OH               Ohio  11727377  11714618  11539336      44826  40860.69   
NY           New York  19223191  19299981  19399878      54555  47126.40   
FL            Florida  22177997  21944577  18845537      65758  53624.76   
DE           Delaware    998619    990334    899593       2489   1948.54   
MD           Maryland   6075314   6065436   5788645      12406   9707.24   
CT        Connecticut   3546588   3552821   3579114       5543   4842.36   
MA      Massachusetts   6922107   6912239   6566307      10554   7800.06   
RI       Rhode Island   1062583   1061509   1053959       1545   1033.81   
NJ         New Jersey   8870685   8874520   8799446       8723   7354.22   

       Density  
State           
AK        1.26  
WY        6.0

That is better, but perhaps we would like to order by densest states first and only show the top five ...
 

In [192]:
states.sort_values(by=['Density'], ascending=False)[:5]

Name  Pop2022  Pop2021  Pop2010  TotalArea  LandArea  Density
State                                                                        
NJ        New Jersey  8870685  8874520  8799446       8723   7354.22  1206.20
RI      Rhode Island  1062583  1061509  1053959       1545   1033.81  1027.83
MA     Massachusetts  6922107  6912239  6566307      10554   7800.06   887.44
CT       Connecticut  3546588  3552821  3579114       5543   4842.36   732.41
MD          Maryland  6075314  6065436  5788645      12406   9707.24   625.85

Finally, let's add a column, growth, that shows the percent change in population from 2021 to 2022.  That formula would be

$$Growth=(\frac{Pop2022}{Pop2021} - 1) \times 100  $$

In [193]:
states['Growth'] = (states['Pop2022'] / states['Pop2021'] - 1) * 100
states[:10]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea  \
State                                                                   
AK          Alaska    720763    724357    713910     665384 570640.95   
AL         Alabama   4949697   4934193   4785437      52420  50645.33   
AR        Arkansas   3042017   3033946   2921964      53179  52035.48   
AZ         Arizona   7640796   7520103   6407172     113990 113594.08   
CA      California  39664128  39613493  37319502     163695 155779.22   
CO        Colorado   5961083   5893634   5047349     104094 103641.89   
CT     Connecticut   3546588   3552821   3579114       5543   4842.36   
DE        Delaware    998619    990334    899593       2489   1948.54   
FL         Florida  22177997  21944577  18845537      65758  53624.76   
GA         Georgia  10936299  10830007   9711881      59425  57513.49   

       Density  Growth  
State                   
AK        1.26   -0.50  
AL       97.73    0.31  
AR       58.46    0.27  
AZ       67.26    1.60  
CA      254.62    0.13  
CO       57.52    1.14  
CT      732.41   -0.18  
DE      512.50    0.84  
FL      413.58    1.06  
GA      190.15    0.98

And let's find the 5 fastest growing states:

In [194]:
states.sort_values(by=['Growth'], ascending=False)[:5]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea  Density  \
State                                                                        
ID       Idaho   1896652   1860123   1570746      83569  82643.12    22.95   
NV      Nevada   3238601   3185786   2702405     110572 109781.18    29.50   
AZ     Arizona   7640796   7520103   6407172     113990 113594.08    67.26   
UT        Utah   3363182   3310774   2775332      84897  82169.62    40.93   
TX       Texas  30097526  29730311  25241971     268596 261231.71   115.21   

       Growth  
State          
ID       1.96  
NV       1.66  
AZ       1.60  
UT       1.58  
TX       1.24


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Descriptive Statistics on DataFrames
One handy function is `describe`

In [195]:
states.describe()

Pop2022     Pop2021     Pop2010  TotalArea  LandArea  Density  \
count       50.00       50.00       50.00      50.00     50.00    50.00   
mean   6643544.68  6612588.28  6174328.80   75932.92  70578.69   204.27   
std    7492630.40  7457726.91  6862379.24   97327.62  85833.06   266.57   
min     582233.00   581075.00   564487.00    1545.00   1033.81     1.26   
25%    1912686.50  1883091.25  1835716.25   37417.00  36741.17    48.00   
50%    4551669.50  4553857.50  4446356.50   57093.50  53891.28   107.48   
75%    7826172.75  7727731.50  6698699.25   84565.00  81225.73   221.41   
max   39664128.00 39613493.00 37319502.00  665384.00 570640.95  1206.20   

       Growth  
count   50.00  
mean     0.44  
std      0.60  
min     -0.69  
25%      0.08  
50%      0.29  
75%      0.83  
max      1.96

Alternatively, we could retrieve a specific statistic:


In [196]:
states['Pop2022'].mean()

6643544.68

or find out how many states lost population in 2022 using `count`:

In [197]:
states.loc[states['Pop2022'] < states['Pop2021']][['Pop2022']].count()

Pop2022    10
dtype: int64

Or get the total population of the United States (minus the District of Columbia)

In [198]:
states['Pop2022'].sum()

332177234

or to make that more readable

In [199]:
totalPop = "{:,}".format(states['Pop2022'].sum())
print('Total Population: ' + totalPop)

Total Population: 332,177,234


What percentage of the U. S. population is accounted for by the top 5 most populous states?

In [200]:
statesByPopulation = states.sort_values(by=['Pop2022'], ascending=False)[:10][['Pop2022']]
percent = (statesByPopulation.sum() / states['Pop2022'].sum()) * 100
print("The top 10 most populous states account for {:4.2f} percent of the U.S. Population" .format(percent['Pop2022']))

The top 10 most populous states account for 54.17 percent of the U.S. Population


## Summary Statistics and Axes

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ax.png)
Suppose I have customers of my vinyl record shop rate different artists

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|5|-|5|2|2|
|Clara|2|-|-|4|5|
|Kelsey|5|5|5|2|-|
|Angelica|2|3|-|5|5|
|Jordyn|2|1|-|5|-|

First we will read in the file ...

In [201]:
ratings = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings.csv', index_col = 0)
ratings

Taylor Swift  Miranda Lambert  Carrie Underwood  Nicki Minaj  \
Customer                                                                 
Jake                 5              NaN              5.00            2   
Clara                2              NaN               NaN            4   
Kelsey               5             5.00              5.00            2   
Angelica             2             3.00               NaN            5   
Jordyn               2             1.00               NaN            5   

          Ariana Grande  
Customer                 
Jake               2.00  
Clara              5.00  
Kelsey              NaN  
Angelica           5.00  
Jordyn              NaN

We can get the mean rating of each artist by:


In [202]:
ratings.mean()

Taylor Swift       3.20
Miranda Lambert    3.00
Carrie Underwood   5.00
Nicki Minaj        3.60
Ariana Grande      4.00
dtype: float64

Note that the summary statistics ignore NaN entries. The mean rating for Miranda Lambert is computed just on the people that rated her.

Many descriptive statistics functions take an optional parameter `axis` that tells which axis to reduce over. If we want the mean ratings for each **customer** instead of each artist we can do:

In [203]:
ratings.mean(axis=1)

Customer
Jake       3.50
Clara      3.67
Kelsey     4.25
Angelica   3.75
Jordyn     2.67
dtype: float64



Sweet! So `axis=1` means reduce by rows and `axis=0` means reduce by columns:

In [204]:
ratings.mean(axis=0)

Taylor Swift       3.20
Miranda Lambert    3.00
Carrie Underwood   5.00
Nicki Minaj        3.60
Ariana Grande      4.00
dtype: float64

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## List of Descriptive Statistics
(from the book *Python for Data Analysis*)

Method | Description
:-- | :--
`count` | Number of non-NaN values
`describe` | A set of common summary statistics
`min, max` | compute minimum and maximum values
`argmin, argmax` | compute index locations of minimum and maximum values
`sum` | Sum the values
`mean` | Mean of values
`median` | Median of values
`std` | Sample standard deviation

So, for example, the lowest rating for each artist:


In [205]:
ratings.min()

Taylor Swift       2.00
Miranda Lambert    1.00
Carrie Underwood   5.00
Nicki Minaj        2.00
Ariana Grande      2.00
dtype: float64


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

# <font color='#EE4C2C'>You Try ...</font> 
Ok, it is time for you to try out what you just learned. Let us start with the electric vehicle datafile we have already seen:

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv
```
When we loaded the datafile before, it worked fine for those examples, but you may encounter an error here, that will require you to modify the `read_csv`.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/evBanner.png)

In [206]:
# TO DO
data = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv', na_values='-',index_col = 'Brand')

data


Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                                           
Tesla         Model 3 Long Range Dual Motor      4.60           233       450   
Volkswagen                        ID.3 Pure     10.00           160       270   
Polestar                                  2      4.70           210       400   
BMW                                     iX3      6.80           180       360   
Honda                                     e      9.50           145       170   
...                                     ...       ...           ...       ...   
Nissan      Ariya e-4ORCE 87kWh Performance      5.10           200       375   
Byton                     M-Byte 95 kWh 2WD      7.50           190       400   
Rivian                                  R1T      3.00           185       505   
GMC                               Hummer EV      3.00           171       529   
Ford                         F150 Lightning      3.00           180       514   

            Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
Brand                                                                
Tesla                   161          940.00         Yes        AWD   
Volkswagen              167          250.00         Yes        RWD   
Polestar                181          620.00         Yes        AWD   
BMW                     206          560.00         Yes        RWD   
Honda                   168          190.00         Yes        RWD   
...                     ...             ...         ...        ...   
Nissan                  232          450.00         Yes        AWD   
Byton                   238          480.00         Yes        AWD   
Rivian                  299          337.00         Yes        AWD   
GMC                     446          250.00         Yes        AWD   
Ford                    299          309.00         Yes        AWD   

              PlugType  BodyStyle Segment  Seats  PriceEuro  
Brand                                                        
Tesla       Type 2 CCS      Sedan       D      5   55480.00  
Volkswagen  Type 2 CCS  Hatchback       C      5   30000.00  
Polestar    Type 2 CCS   Liftback       D      5   56440.00  
BMW         Type 2 CCS        SUV       D      5   68040.00  
Honda       Type 2 CCS  Hatchback       B      4   32997.00  
...                ...        ...     ...    ...        ...  
Nissan      Type 2 CCS  Hatchback       C      5   65000.00  
Byton       Type 2 CCS        SUV       E      5   62000.00  
Rivian      Type 2 CCS     Pickup       N      6   75000.00  
GMC         Type 2 CCS      Truck       E      5        NaN  
Ford        Type 2 CCS      Truck       E      5        NaN  

[106 rows x 13 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

# <font color='#EE4C2C'>You Try ...</font> 
Ok, it is time for you to try out what you just learned. Let us start with the electric vehicle datafile we have already seen:

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv
```
When we loaded the datafile before, it worked fine for those examples, but you may encounter an error here, that will require you to modify the `read_csv`.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/evBanner.png)

Some of the columns are based on kilometers:

* TopSpeed_KmH	
* Range_Km	
* Efficiency_WhKm	
* FastCharge_KmH

and we are first going to add USA mile-centric columns.

## <font color='#EE4C2C'>1. TopSpeed_MPH</font>

Add a new column `TopSpeed_MPH` computed from the `TopSpeed_KmH` column, and just to check, display the first 10 rows. (The Lucid Air should have a top speed slightly over 155 MPH)


In [207]:
data['TopSpeed_MPH'] = data['TopSpeed_KmH'] * .621



## <font color='#EE4C2C'>2. Range and FastCharge_MPH</font>

Add a new column `Range` computed from the `Range_Km` column that shows the range in miles.

Also add a column `FastCharge_MPH computed from `FastCharge_KmH`. This will indicate how many miles can you go on one hour of charging. 

Just to check, display the first 10 rows. (The Tesla Model 3 Long Range Dual Motor should have a range of nearly 280 miles and charge at a rate of 584 miles per hour of fast charging.)

In [208]:
data['Range_Mi'] = data['Range_Km'] * .621
data['FastCharge_MPH'] = data['FastCharge_KmH'] * .621
data[:10]

Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                                         
Tesla       Model 3 Long Range Dual Motor      4.60           233       450   
Volkswagen                      ID.3 Pure     10.00           160       270   
Polestar                                2      4.70           210       400   
BMW                                   iX3      6.80           180       360   
Honda                                   e      9.50           145       170   
Lucid                                 Air      2.80           250       610   
Volkswagen                         e-Golf      9.60           150       190   
Peugeot                             e-208      8.10           150       275   
Tesla         Model 3 Standard Range Plus      5.60           225       310   
Audi                            Q4 e-tron      6.30           180       400   

            Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
Brand                                                                
Tesla                   161          940.00         Yes        AWD   
Volkswagen              167          250.00         Yes        RWD   
Polestar                181          620.00         Yes        AWD   
BMW                     206          560.00         Yes        RWD   
Honda                   168          190.00         Yes        RWD   
Lucid                   180          620.00         Yes        AWD   
Volkswagen              168          220.00         Yes        FWD   
Peugeot                 164          420.00         Yes        FWD   
Tesla                   153          650.00         Yes        RWD   
Audi                    193          540.00         Yes        AWD   

              PlugType  BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  \
Brand                                                                       
Tesla       Type 2 CCS      Sedan       D      5   55480.00        144.69   
Volkswagen  Type 2 CCS  Hatchback       C      5   30000.00         99.36   
Polestar    Type 2 CCS   Liftback       D      5   56440.00        130.41   
BMW         Type 2 CCS        SUV       D      5   68040.00        111.78   
Honda       Type 2 CCS  Hatchback       B      4   32997.00         90.05   
Lucid       Type 2 CCS      Sedan       F      5  105000.00        155.25   
Volkswagen  Type 2 CCS  Hatchback       C      5   31900.00         93.15   
Peugeot     Type 2 CCS  Hatchback       B      5   29682.00         93.15   
Tesla       Type 2 CCS      Sedan       D      5   46380.00        139.72   
Audi        Type 2 CCS        SUV       D      5   55000.00        111.78   

            Range_Mi  FastCharge_MPH  
Brand                                 
Tesla         279.45          583.74  
Volkswagen    167.67          155.25  
Polestar      248.40          385.02  
BMW           223.56          347.76  
Honda         105.57          117.99  
Lucid         378.81          385.02  
Volkswagen    117.99          136.62  
Peugeot       170.78          260.82  
Tesla         192.51          403.65  
Audi          248.40          335.34

## <font color='#EE4C2C'>3. Efficiency_MkwH</font>

The DataFrame has a column `Efficiency_WhKm` which indicates how many watt hours does it take to go one kilometer. One common measure of efficieny is how many miles can you go on one kilowatt hour (kwH). So please add this column to the DataFrame. (The Tesla Model 3 Standard Range Plus should have an efficiency of around 4 miles per kilowatt hour.) 


In [209]:
data['Efficiency_MkwH'] = 1000 / 1.609344 / data['Efficiency_WhKm']
data

Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                                           
Tesla         Model 3 Long Range Dual Motor      4.60           233       450   
Volkswagen                        ID.3 Pure     10.00           160       270   
Polestar                                  2      4.70           210       400   
BMW                                     iX3      6.80           180       360   
Honda                                     e      9.50           145       170   
...                                     ...       ...           ...       ...   
Nissan      Ariya e-4ORCE 87kWh Performance      5.10           200       375   
Byton                     M-Byte 95 kWh 2WD      7.50           190       400   
Rivian                                  R1T      3.00           185       505   
GMC                               Hummer EV      3.00           171       529   
Ford                         F150 Lightning      3.00           180       514   

            Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
Brand                                                                
Tesla                   161          940.00         Yes        AWD   
Volkswagen              167          250.00         Yes        RWD   
Polestar                181          620.00         Yes        AWD   
BMW                     206          560.00         Yes        RWD   
Honda                   168          190.00         Yes        RWD   
...                     ...             ...         ...        ...   
Nissan                  232          450.00         Yes        AWD   
Byton                   238          480.00         Yes        AWD   
Rivian                  299          337.00         Yes        AWD   
GMC                     446          250.00         Yes        AWD   
Ford                    299          309.00         Yes        AWD   

              PlugType  BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  \
Brand                                                                       
Tesla       Type 2 CCS      Sedan       D      5   55480.00        144.69   
Volkswagen  Type 2 CCS  Hatchback       C      5   30000.00         99.36   
Polestar    Type 2 CCS   Liftback       D      5   56440.00        130.41   
BMW         Type 2 CCS        SUV       D      5   68040.00        111.78   
Honda       Type 2 CCS  Hatchback       B      4   32997.00         90.05   
...                ...        ...     ...    ...        ...           ...   
Nissan      Type 2 CCS  Hatchback       C      5   65000.00        124.20   
Byton       Type 2 CCS        SUV       E      5   62000.00        117.99   
Rivian      Type 2 CCS     Pickup       N      6   75000.00        114.89   
GMC         Type 2 CCS      Truck       E      5        NaN        106.19   
Ford        Type 2 CCS      Truck       E      5        NaN        111.78   

            Range_Mi  FastCharge_MPH  Efficiency_MkwH  
Brand                                                  
Tesla         279.45          583.74             3.86  
Volkswagen    167.67          155.25             3.72  
Polestar      248.40          385.02             3.43  
BMW           223.56          347.76             3.02  
Honda         105.57          117.99             3.70  
...              ...             ...              ...  
Nissan        232.88          279.45             2.68  
Byton         248.40          298.08             2.61  
Rivian        313.61          209.28             2.08  
GMC           328.51          155.25             1.39  
Ford          319.19          191.89             2.08  

[106 rows x 17 columns]

## <font color='#EE4C2C'>5. Most and least efficient</font>
Using the bevs DataFrame, what are the 5 most efficient evs ordered by the most efficient first?


In [210]:
mostEf = data.sort_values(by = 'Efficiency_MkwH', ascending = False)[:5]
mostEf

Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                                      
Lightyear                          One     10.00           150       575   
Tesla      Model 3 Standard Range Plus      5.60           225       310   
Hyundai                 IONIQ Electric      9.70           165       250   
Hyundai           Kona Electric 39 kWh      9.90           155       255   
Mini                         Cooper SE      7.30           150       185   

           Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
Brand                                                                           
Lightyear              104          540.00         Yes        AWD  Type 2 CCS   
Tesla                  153          650.00         Yes        RWD  Type 2 CCS   
Hyundai                153          210.00         Yes        FWD  Type 2 CCS   
Hyundai                154          210.00         Yes        FWD  Type 2 CCS   
Mini                   156          260.00         Yes        FWD  Type 2 CCS   

           BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  Range_Mi  \
Brand                                                                    
Lightyear   Liftback       F      5  149000.00         93.15    357.07   
Tesla          Sedan       D      5   46380.00        139.72    192.51   
Hyundai     Liftback       C      5   34459.00        102.47    155.25   
Hyundai          SUV       B      5   33971.00         96.25    158.35   
Mini       Hatchback       B      4   31681.00         93.15    114.89   

           FastCharge_MPH  Efficiency_MkwH  
Brand                                       
Lightyear          335.34             5.97  
Tesla              403.65             4.06  
Hyundai            130.41             4.06  
Hyundai            130.41             4.03  
Mini               161.46             3.98

What are the five most inefficient evs ordered by least efficient first?

In [211]:
leastEF = data.sort_values(by = 'Efficiency_MkwH')[:5]
leastEF

Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                             
GMC                 Hummer EV      3.00           171       529   
Ford           F150 Lightning      3.00           180       514   
Rivian                    R1T      3.00           185       505   
Mercedes         EQV 300 Long     10.00           140       330   
Audi      e-tron S 55 quattro      4.50           210       320   

          Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
Brand                                                                          
GMC                   446          250.00         Yes        AWD  Type 2 CCS   
Ford                  299          309.00         Yes        AWD  Type 2 CCS   
Rivian                299          337.00         Yes        AWD  Type 2 CCS   
Mercedes              273          290.00         Yes        FWD  Type 2 CCS   
Audi                  270          510.00         Yes        AWD  Type 2 CCS   

         BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  Range_Mi  \
Brand                                                                  
GMC          Truck       E      5        NaN        106.19    328.51   
Ford         Truck       E      5        NaN        111.78    319.19   
Rivian      Pickup       N      6   75000.00        114.89    313.61   
Mercedes       SPV       N      7   70631.00         86.94    204.93   
Audi           SUV       E      5   93800.00        130.41    198.72   

          FastCharge_MPH  Efficiency_MkwH  
Brand                                      
GMC               155.25             1.39  
Ford              191.89             2.08  
Rivian            209.28             2.08  
Mercedes          180.09             2.28  
Audi              316.71             2.30

## <font color='#EE4C2C'>6. fastest
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/porsche.png)

The Porsche 911 997 Turbo S is among the world's fastest production combustion cars with a 0-60 acceleration of 2.9 seconds. (In our dataset `AccelSec` measures the same thing.) Which electric cars have better acceleration than this? 

In [212]:
fastAccel = data.loc[data['AccelSec'] < 2.9]
fastAccel

Model  AccelSec  TopSpeed_KmH  Range_Km  \
Brand                                                            
Lucid                    Air      2.80           250       610   
Porsche       Taycan Turbo S      2.80           260       375   
Tesla               Roadster      2.10           410       970   
Tesla    Model S Performance      2.50           261       505   
Tesla    Model X Performance      2.80           250       440   

         Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
Brand                                                                         
Lucid                180          620.00         Yes        AWD  Type 2 CCS   
Porsche              223          780.00         Yes        AWD  Type 2 CCS   
Tesla                206          920.00         Yes        AWD  Type 2 CCS   
Tesla                188          550.00         Yes        AWD      Type 2   
Tesla                216          480.00         Yes        AWD      Type 2   

        BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  Range_Mi  \
Brand                                                                 
Lucid       Sedan       F      5  105000.00        155.25    378.81   
Porsche     Sedan       F      4  180781.00        161.46    232.88   
Tesla      Cabrio       S      4  215000.00        254.61    602.37   
Tesla    Liftback       F      5   96990.00        162.08    313.61   
Tesla         SUV       F      7  102990.00        155.25    273.24   

         FastCharge_MPH  Efficiency_MkwH  
Brand                                     
Lucid            385.02             3.45  
Porsche          484.38             2.79  
Tesla            571.32             3.02  
Tesla            341.55             3.31  
Tesla            298.08             2.88

## <font color='#EE4C2C'>7. Refrigerators
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/frigTesla.png)

Representative Thomas Massie (R-KY) at a hearing questioning Secretarty Pete Buttigieg says the energy (kWh) required to home charge an electric vehicle for a year is 25 times greater than running the average refrigerator. He also said "It would take four times as much electricity to charge the average household's cars as the average household uses on air conditioning." And he said, "that means the average household would use twice as much electricity charging one of their cars as they would use for all of the air conditioning that they use for the entire year." Is this true? 

Since air conditioner use is quite varied in different regions of the country, perhaps a better measure would be the energy used for air conditioning and heating combined.

You are going to need to do some web searching to find the typical energy use of household these household items.

What do you think about the usefulness of this comparing household appliances to EVs when making government policy?  


<h1> TESLAS V FRIDGES</h1>

1.   An average fridge will use 350 kwh/year
2.   An average call will be driven 15000 miles per year. Assuming an efficiency of .34 kwh per mile, that means 5100 kwh per year.

A tesla will use much more energy per year, but the comparison is flawed. Trying to reduce an almost performance level vehicle that moves thousands of pounds, that happens to be electric, and compare it to a refrigerator in an effort to guide public policy by saying one uses more power than the other is stupid.

## <font color='#EE4C2C'>8. Global Adoption of EVs

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/earth.png)

The data file we are using for this is

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv
```

This data is from the [International Energy Agency](https://www.iea.org/data-and-statistics/data-product/global-ev-outlook-2022).

The first step in data mining work is to examine the dataset, which often is not in the best format to directly answer our questions. That is the case with this file. You have the Pandas skills, but you will need to spend a bit of time looking at the datafile and understanding it, so you can write some code to meet the need.

So what is our need? We are interested in what percent of new car sales are evs in 2021. We are interested in historical data, not predictions. We would like to see the top five countries with the best adoption percent and the bottom five. Have appropriate labels indicating the list of top 5 and bottom 5. 



In [213]:
##getting and cleaning data
evdata = pd.read_csv("https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv", na_values = '-')
evdata

region          category              parameter    mode powertrain  \
0     Australia        Historical               EV sales    Cars        BEV   
1     Australia        Historical               EV stock    Cars        BEV   
2     Australia        Historical         EV sales share    Cars         EV   
3     Australia        Historical         EV stock share    Cars         EV   
4     Australia        Historical         EV stock share    Cars         EV   
...         ...               ...                    ...     ...        ...   
7011      World  Projection-STEPS   Oil displacement Mbd    Cars         EV   
7012      World  Projection-STEPS  Oil displacement Mlge   Buses         EV   
7013      World  Projection-STEPS  Oil displacement Mlge  Trucks         EV   
7014      World  Projection-STEPS  Oil displacement Mlge    Vans         EV   
7015      World  Projection-STEPS  Oil displacement Mlge    Cars         EV   

      year                               unit     value  
0     2011                              sales     49.00  
1     2011                              stock     49.00  
2     2011                            percent      0.01  
3     2011                            percent      0.00  
4     2012                            percent      0.00  
...    ...                                ...       ...  
7011  2030             Milion barrels per day      2.41  
7012  2030  Milion litres gasoline equivalent  30311.88  
7013  2030  Milion litres gasoline equivalent  13943.43  
7014  2030  Milion litres gasoline equivalent  16207.84  
7015  2030  Milion litres gasoline equivalent 141228.52  

[7016 rows x 8 columns]

In [214]:
#the following gets the data that: is historical, has a year of 2021 and is an EV sale
historicaldata = evdata[(evdata['category'] == 'Historical') & (evdata['parameter'] == 'EV sales') & 
                        (evdata['mode'] == 'Cars') & (evdata['year'] == 2021) & (evdata['region'] != 'World')]
historicaldata

region    category parameter  mode powertrain  year   unit  \
68         Australia  Historical  EV sales  Cars       PHEV  2021  sales   
70         Australia  Historical  EV sales  Cars        BEV  2021  sales   
282          Belgium  Historical  EV sales  Cars       PHEV  2021  sales   
285          Belgium  Historical  EV sales  Cars        BEV  2021  sales   
355           Brazil  Historical  EV sales  Cars        BEV  2021  sales   
356           Brazil  Historical  EV sales  Cars       PHEV  2021  sales   
480           Canada  Historical  EV sales  Cars       PHEV  2021  sales   
489           Canada  Historical  EV sales  Cars        BEV  2021  sales   
574            Chile  Historical  EV sales  Cars       PHEV  2021  sales   
576            Chile  Historical  EV sales  Cars        BEV  2021  sales   
1040           China  Historical  EV sales  Cars       PHEV  2021  sales   
1042           China  Historical  EV sales  Cars        BEV  2021  sales   
1562         Denmark  Historical  EV sales  Cars       PHEV  2021  sales   
1565         Denmark  Historical  EV sales  Cars        BEV  2021  sales   
2299         Finland  Historical  EV sales  Cars        BEV  2021  sales   
2301         Finland  Historical  EV sales  Cars       PHEV  2021  sales   
2539          France  Historical  EV sales  Cars       PHEV  2021  sales   
2540          France  Historical  EV sales  Cars        BEV  2021  sales   
2757         Germany  Historical  EV sales  Cars        BEV  2021  sales   
2758         Germany  Historical  EV sales  Cars       PHEV  2021  sales   
2840          Greece  Historical  EV sales  Cars        BEV  2021  sales   
2844          Greece  Historical  EV sales  Cars       PHEV  2021  sales   
3009         Iceland  Historical  EV sales  Cars       PHEV  2021  sales   
3017         Iceland  Historical  EV sales  Cars        BEV  2021  sales   
3135           India  Historical  EV sales  Cars        BEV  2021  sales   
3137           India  Historical  EV sales  Cars       PHEV  2021  sales   
3516           Italy  Historical  EV sales  Cars       PHEV  2021  sales   
3519           Italy  Historical  EV sales  Cars        BEV  2021  sales   
3695           Japan  Historical  EV sales  Cars       PHEV  2021  sales   
3697           Japan  Historical  EV sales  Cars        BEV  2021  sales   
3842           Korea  Historical  EV sales  Cars        BEV  2021  sales   
3848           Korea  Historical  EV sales  Cars       PHEV  2021  sales   
3915          Mexico  Historical  EV sales  Cars        BEV  2021  sales   
3920          Mexico  Historical  EV sales  Cars       PHEV  2021  sales   
4153     Netherlands  Historical  EV sales  Cars       PHEV  2021  sales   
4157     Netherlands  Historical  EV sales  Cars        BEV  2021  sales   
4288     New Zealand  Historical  EV sales  Cars        BEV  2021  sales   
4295     New Zealand  Historical  EV sales  Cars       PHEV  2021  sales   
4480          Norway  Historical  EV sales  Cars        BEV  2021  sales   
4494          Norway  Historical  EV sales  Cars       PHEV  2021  sales   
4711    Other Europe  Historical  EV sales  Cars        BEV  2021  sales   
4712    Other Europe  Historical  EV sales  Cars       PHEV  2021  sales   
4842          Poland  Historical  EV sales  Cars       PHEV  2021  sales   
4844          Poland  Historical  EV sales  Cars        BEV  2021  sales   
4981        Portugal  Historical  EV sales  Cars        BEV  2021  sales   
4983        Portugal  Historical  EV sales  Cars       PHEV  2021  sales   
5501    South Africa  Historical  EV sales  Cars       PHEV  2021  sales   
5503    South Africa  Historical  EV sales  Cars        BEV  2021  sales   
5677           Spain  Historical  EV sales  Cars       PHEV  2021  sales   
5686           Spain  Historical  EV sales  Cars        BEV  2021  sales   
5879          Sweden  Historical  EV sales  Cars       PHEV  2021  sales   
5882          Sweden  Historical  EV sales  Cars        BEV  2

In [215]:
##USING PIVOT TABLE TO SUM SALES VALUES
##I was going to solve this problem by extracking all regions into an array using np, 
##then removing duplicates and use each individual country to iterate and sum over the historicaldata dataframe.
##But I thought to myself there must be a better way to do this given pandas' capabilites.
##I remembered a DataFrame was like an excel spreadsheet, and spreadsheets have a function called pivot tables.
##At that point it was just a matter of finding the correct function in pandas.
aggdata = historicaldata.pivot_table(values = 'value', index = ['region'], aggfunc = np.sum)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
aggdata


value
region                   
Australia        20700.00
Belgium          70680.00
Brazil            7254.00
Canada           87700.00
Chile              566.00
China          3334161.00
Denmark          65650.00
Finland          30314.00
France          308663.00
Germany         681410.00
Greece            6961.00
Iceland           9149.00
India            11739.00
Italy           138012.00
Japan            44296.00
Korea            90546.00
Mexico            4290.00
Netherlands      93924.00
New Zealand      10300.00
Norway          151892.00
Other Europe    109776.00
Poland           17842.00
Portugal         29182.00
South Africa       269.00
Spain            67721.00
Sweden          136081.00
Switzerland      53489.00
USA             631152.00
United Kingdom  312276.00

In [216]:
#top5
top5 = aggdata.sort_values(by = 'value', ascending = False)[:5]
top5

value
region                   
China          3334161.00
Germany         681410.00
USA             631152.00
United Kingdom  312276.00
France          308663.00

In [217]:
#bottom5
bottom5 = aggdata.sort_values(by = 'value', ascending = True)[:5]
bottom5

value
region              
South Africa  269.00
Chile         566.00
Mexico       4290.00
Greece       6961.00
Brazil       7254.00

## <font color='#EE4C2C'>9. Adoption of EVs Worldwide Part 2
In the U.S. in 2021 Electric Vehicles account for 2.5% of total new car sales. Is this higher or lower than the average (mean) 2021 figures in our dataset?

Also, how many countries have a higher adoption rate than the U.S?

In addition to cells showing your work, please have some readable content.

In [218]:
##getting all evsales shares into one dataframe
evsalesshare2021 = evdata[(evdata['category'] == 'Historical') & (evdata['parameter'] == 'EV sales share') & (evdata['mode'] == 'Cars') & (evdata['year'] == 2021) & (evdata['region'] != 'World')]
evsalesshare2021

region    category       parameter  mode powertrain  year  \
73         Australia  Historical  EV sales share  Cars         EV  2021   
287          Belgium  Historical  EV sales share  Cars         EV  2021   
362           Brazil  Historical  EV sales share  Cars         EV  2021   
484           Canada  Historical  EV sales share  Cars         EV  2021   
579            Chile  Historical  EV sales share  Cars         EV  2021   
1092           China  Historical  EV sales share  Cars         EV  2021   
1572         Denmark  Historical  EV sales share  Cars         EV  2021   
1896          Europe  Historical  EV sales share  Cars         EV  2021   
2311         Finland  Historical  EV sales share  Cars         EV  2021   
2546          France  Historical  EV sales share  Cars         EV  2021   
2749         Germany  Historical  EV sales share  Cars         EV  2021   
2842          Greece  Historical  EV sales share  Cars         EV  2021   
3012         Iceland  Historical  EV sales share  Cars         EV  2021   
3139           India  Historical  EV sales share  Cars         EV  2021   
3513           Italy  Historical  EV sales share  Cars         EV  2021   
3689           Japan  Historical  EV sales share  Cars         EV  2021   
3839           Korea  Historical  EV sales share  Cars         EV  2021   
3916          Mexico  Historical  EV sales share  Cars         EV  2021   
4139     Netherlands  Historical  EV sales share  Cars         EV  2021   
4285     New Zealand  Historical  EV sales share  Cars         EV  2021   
4490          Norway  Historical  EV sales share  Cars         EV  2021   
4705    Other Europe  Historical  EV sales share  Cars         EV  2021   
4853          Poland  Historical  EV sales share  Cars         EV  2021   
4992        Portugal  Historical  EV sales share  Cars         EV  2021   
5498    South Africa  Historical  EV sales share  Cars         EV  2021   
5670           Spain  Historical  EV sales share  Cars         EV  2021   
5873          Sweden  Historical  EV sales share  Cars         EV  2021   
6049     Switzerland  Historical  EV sales share  Cars         EV  2021   
6308  United Kingdom  Historical  EV sales share  Cars         EV  2021   
6486             USA  Historical  EV sales share  Cars         EV  2021   

         unit  value  
73    percent   2.82  
287   percent  18.45  
362   percent   0.47  
484   percent   6.60  
579   percent   0.19  
1092  percent  16.01  
1572  percent  35.42  
1896  percent  16.62  
2311  percent  30.78  
2546  percent  18.90  
2749  percent  25.99  
2842  percent   6.90  
3012  percent  71.73  
3139  percent   0.38  
3513  percent   9.47  
3689  percent   1.00  
3839  percent   6.24  
3916  percent   0.50  
4139  percent  29.73  
4285  percent   4.40  
4490  percent  86.17  
4705  percent   4.89  
4853  percent   1.36  
4992  percent  19.90  
5498  percent   0.09  
5670  percent   7.88  
5873  percent  43.29  
6049  percent  22.43  
6308  percent  18.96  
6486  percent   4.56

In [219]:
share2021pivot = evsalesshare2021.pivot_table(values = 'value', index = ['region'], aggfunc = np.sum)
share2021pivot

value
region               
Australia        2.82
Belgium         18.45
Brazil           0.47
Canada           6.60
Chile            0.19
China           16.01
Denmark         35.42
Europe          16.62
Finland         30.78
France          18.90
Germany         25.99
Greece           6.90
Iceland         71.73
India            0.38
Italy            9.47
Japan            1.00
Korea            6.24
Mexico           0.50
Netherlands     29.73
New Zealand      4.40
Norway          86.17
Other Europe     4.89
Poland           1.36
Portugal        19.90
South Africa     0.09
Spain            7.88
Sweden          43.29
Switzerland     22.43
USA              4.56
United Kingdom  18.96

In [220]:
##according to this data, EV Car sales account for 4.56 percent of all new car sales
##getting mean of all countries
meansales = share2021pivot['value'].mean()
meansales

17.07053746233384

The world mean of ev sales as a percentage of new car sales is 17.07%. In the US this figure is 4.56%, meaning much lower.

In [237]:
##finding countries with means higher than 4.56%
higherAdoptionCountries = share2021pivot[(share2021pivot['value'] > 4.57)]
higherAdoptionCountries


value
region               
Belgium         18.45
Canada           6.60
China           16.01
Denmark         35.42
Europe          16.62
Finland         30.78
France          18.90
Germany         25.99
Greece           6.90
Iceland         71.73
Italy            9.47
Korea            6.24
Netherlands     29.73
Norway          86.17
Other Europe     4.89
Portugal        19.90
Spain            7.88
Sweden          43.29
Switzerland     22.43
United Kingdom  18.96

In [238]:
##getting number of countries that are greater than 4.56
print(len(higherAdoptionCountries))

20


According to my calculations, 20 countries have a greater ev sales percentage than the United states.